### PR TITLE
I-121 Add diff implementation

### LINKED ISSUE #121: k8s-handle diff support
Hello,
How hard is to add `kubectl diff` functionality right into k8s-handle?
It is handy to have a test for Merge Request actually showing what would be changed in Cluster after all templates are rendered etc.
Right now we are using something like:
```
k8s-handle render -s section

kubectl config set-credentials user --token="$K8S_TOKEN"
kubectl config set-cluster cluster --server=https://k8s --insecure-skip-tls-verify=true
kubectl config set-context context --cluster=cluster --user=user
kubectl config use-context context

kubectl diff -n namespace -R -f /tmp/k8s-handle
```
While running steps in docker, this requires container with both k8s-handle and kubectl - which complicates stuff.
Advantage to have diff in python is that it's format would be easier to customize (drop unnecessary fields etc)

Another command where knowledge about objects in the Cluster could be useful is `deploy`.
Because for new users all these logs looks scary:
`INFO:k8s_handle.k8s.provisioner:Deployment "prometheus" already exists, replace it`
Why do you replace the object if it does not changed? For example `kubectl apply` shows if object was actually changed or not. Would be great to show diff on `deploy` too, or skip objects which has zero diff.

What do you think?

### PR DESCRIPTION
**Pull request by rvadim**
Fixes #121 
Very basic and dirty implementation of diff command. Can be improved in future.

### CODE REVIEW

File: k8s_handle/k8s/diff.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+log = logging.getLogger(__name__)
+
+IGNORE_FIELDS = [
+    'metadata.annotations:kubectl.kubernetes.io/last-applied-configuration',
+    'metadata.annotations:deployment.kubernetes.io/revision',
+    'metadata:creationTimestamp',
+    'metadata:resourceVersion',
+    'metadata:selfLink',
+    'metadata:uid',
+    'metadata:namespace',
```
Comment (dekhtyarev) [inline k8s_handle/k8s/diff.py]: Very often, this value can be overridden. Can leave it to compare?

**Comment by dekhtyarev** [comment]
@rvadim, thanks for the PR!
Several times already thought about the diff in k8s-handle

**Comment by seleznev** [comment]
Looks like it doesn't work well for CRD with openAPIV3Schema.
<details>
<summary>how to reproduce</summary>

* template:
  ```yaml
  apiVersion: apiextensions.k8s.io/v1beta1
  kind: CustomResourceDefinition
  metadata:
    name: tests.test.example.com
  spec:
    group: test.example.com
    names:
      kind: Test
      plural: tests
    scope: Namespaced
    version: v1alpha1
    validation:
      openAPIV3Schema:
        properties:
          apiVersion:
            description: 'test description'
            type: string
          metadata:
            type: object
  ```
* diff before first deploy:
  ```
  # k8s-handle diff -s testing-n3
  2020-03-23 05:35:51 INFO:k8s_handle.templating:Trying to generate file from template "cert-manager/customresourcedefinition.yaml.j2" in "/tmp/k8s-handle"
  2020-03-23 05:35:51 INFO:k8s_handle.templating:File "/tmp/k8s-handle/cert-manager/customresourcedefinition.yaml" successfully generated
  2020-03-23 05:35:51 INFO:k8s_handle:Default namespace "kube-system"
  2020-03-23 05:35:52 WARNING:k8s_handle.k8s.deprecation_checker:
          ▄▄
         ████
        ██▀▀██
       ███  ███     Version apiextensions.k8s.io/v1beta1
      ████▄▄████    for resource type CustomResourceDefinition
     █████  █████   is deprecated since 1.16.0
    ██████████████
                    
  2020-03-23 05:35:52 INFO:k8s_handle.k8s.diff: Kind: "CustomResourceDefinition", name: "tests.test.example.com"
  - {}
  + apiVersion: apiextensions.k8s.io/v1beta1
  + kind: CustomResourceDefinition
  + metadata:
  +   name: tests.test.example.com
  + spec:
  +   group: test.example.com
  +   names:
  +     kind: Test
  +     plural: tests
  +   scope: Namespaced
  +   validation:
  +     openAPIV3Schema:
  +       properties:
  +         apiVersion:
  +           description: test description
  +           type: string
  +         metadata:
  +           type: object
  +   version: v1alpha1

                           _(_)_                          wWWWw   _
               @@@@       (_)@(_)   vVVVv     _     @@@@  (___) _(_)_
              @@()@@ wWWWw  (_)\    (___)   _(_)_  @@()@@   Y  (_)@(_)
               @@@@  (___)     `|/    Y    (_)@(_)  @@@@   \|/   (_)
                /      Y       \|    \|/    /(_)    \|      |/      |
             \ |     \ |/       | / \ | /  \|/       |/    \|      \|/
              \|//    \|///    \|//  \|/// \|///    \|//    |//    \|//
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ```
* deploy:
  ```
  # k8s-handle deploy -s testing-n3
  2020-03-23 05:36:07 INFO:k8s_handle.templating:Trying to generate file from template "cert-manager/customresourcedefinition.yaml.j2" in "/tmp/k8s-handle"
  2020-03-23 05:36:07 INFO:k8s_handle.templating:File "/tmp/k8s-handle/cert-manager/customresourcedefinition.yaml" successfully generated
  2020-03-23 05:36:07 INFO:k8s_handle:Default namespace "kube-system"
  2020-03-23 05:36:07 WARNING:k8s_handle.k8s.deprecation_checker:
          ▄▄
         ████
        ██▀▀██
       ███  ███     Version apiextensions.k8s.io/v1beta1
      ████▄▄████    for resource type CustomResourceDefinition
     █████  █████   is deprecated since 1.16.0
    ██████████████
                    
  2020-03-23 05:36:07 INFO:k8s_handle.k8s.provisioner:Using namespace "kube-system"
  2020-03-23 05:36:07 INFO:k8s_handle.k8s.provisioner:CustomResourceDefinition "tests.test.example.com" does not exist, create it
  2020-03-23 05:36:07 ERROR:k8s_handle.k8s.adapters:Invalid value for `conditions`, must not be `None`

                           _(_)_                          wWWWw   _
               @@@@       (_)@(_)   vVVVv     _     @@@@  (___) _(_)_
              @@()@@ wWWWw  (_)\    (___)   _(_)_  @@()@@   Y  (_)@(_)
               @@@@  (___)     `|/    Y    (_)@(_)  @@@@   \|/   (_)
                /      Y       \|    \|/    /(_)    \|      |/      |
             \ |     \ |/       | / \ | /  \|/       |/    \|      \|/
              \|//    \|///    \|//  \|/// \|///    \|//    |//    \|//
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ```
* diff after deploy:
  ```
  # k8s-handle diff -s testing-n3
  2020-03-23 05:36:17 INFO:k8s_handle.templating:Trying to generate file from template "cert-manager/customresourcedefinition.yaml.j2" in "/tmp/k8s-handle"
  2020-03-23 05:36:17 INFO:k8s_handle.templating:File "/tmp/k8s-handle/cert-manager/customresourcedefinition.yaml" successfully generated
  2020-03-23 05:36:17 INFO:k8s_handle:Default namespace "kube-system"
  2020-03-23 05:36:17 WARNING:k8s_handle.k8s.deprecation_checker:
          ▄▄
         ████
        ██▀▀██
       ███  ███     Version apiextensions.k8s.io/v1beta1
      ████▄▄████    for resource type CustomResourceDefinition
     █████  █████   is deprecated since 1.16.0
    ██████████████
                    
  Traceback (most recent call last):
    File "/usr/local/bin/k8s-handle", line 11, in <module>
      load_entry_point('k8s-handle==0.0.0', 'console_scripts', 'k8s-handle')()
    File "/usr/local/lib/python3.6/site-packages/k8s_handle-0.0.0-py3.6.egg/k8s_handle/__init__.py", line 241, in main
      args.func(args_dict)
    File "/usr/local/lib/python3.6/site-packages/k8s_handle-0.0.0-py3.6.egg/k8s_handle/__init__.py", line 55, in handler_diff
      _handler_deploy_destroy(args, COMMAND_DIFF)
    File "/usr/local/lib/python3.6/site-packages/k8s_handle-0.0.0-py3.6.egg/k8s_handle/__init__.py", line 75, in _handler_deploy_destroy
      args.get('show_logs')
    File "/usr/local/lib/python3.6/site-packages/k8s_handle-0.0.0-py3.6.egg/k8s_handle/__init__.py", line 130, in _handler_provision
      executor.run(resource)
    File "/usr/local/lib/python3.6/site-packages/k8s_handle-0.0.0-py3.6.egg/k8s_handle/k8s/diff.py", line 75, in run
      current = yaml.safe_dump(current_dict)
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/__init__.py", line 306, in safe_dump
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/__init__.py", line 278, in dump_all
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 27, in represent
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 48, in represent_data
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 207, in represent_dict
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 118, in represent_mapping
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 48, in represent_data
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 207, in represent_dict
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 118, in represent_mapping
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 48, in represent_data
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 207, in represent_dict
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 118, in represent_mapping
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 48, in represent_data
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 207, in represent_dict
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 118, in represent_mapping
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 48, in represent_data
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 207, in represent_dict
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 118, in represent_mapping
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 58, in represent_data
    File "/usr/local/lib/python3.6/site-packages/PyYAML-5.1-py3.6-linux-x86_64.egg/yaml/representer.py", line 231, in represent_undefined
  yaml.representer.RepresenterError: ('cannot represent an object', {'_not': None,
   'additional_items': None,
   'additional_properties': None,
   'all_of': None,
   'any_of': None,
   'default': None,
   'definitions': None,
   'dependencies': None,
   'description': 'test description',
   'enum': None,
   'example': None,
   'exclusive_maximum': None,
   'exclusive_minimum': None,
   'external_docs': None,
   'format': None,
   'id': None,
   'items': None,
   'max_items': None,
   'max_length': None,
   'max_properties': None,
   'maximum': None,
   'min_items': None,
   'min_length': None,
   'min_properties': None,
   'minimum': None,
   'multiple_of': None,
   'nullable': None,
   'one_of': None,
   'pattern': None,
   'pattern_properties': None,
   'properties': None,
   'ref': None,
   'required': None,
   'schema': None,
   'title': None,
   'type': 'string',
   'unique_items': None})
  ```
* yaml by `kubectl get crd tests.test.example.com -o yaml`:
  ```yaml
  apiVersion: apiextensions.k8s.io/v1
  kind: CustomResourceDefinition
  metadata:
    creationTimestamp: "2020-03-23T05:36:07Z"
    generation: 1
    name: tests.test.example.com
    resourceVersion: "184115001"
    selfLink: /apis/apiextensions.k8s.io/v1/customresourcedefinitions/tests.test.example.com
    uid: 7a942be1-4852-4712-9fe2-70fbbd3fd4d4
  spec:
    conversion:
      strategy: None
    group: test.example.com
    names:
      kind: Test
      listKind: TestList
      plural: tests
      singular: test
    preserveUnknownFields: true
    scope: Namespaced
    versions:
    - name: v1alpha1
      schema:
        openAPIV3Schema:
          properties:
            apiVersion:
              description: test description
              type: string
            metadata:
              type: object
      served: true
      storage: true
  status:
    acceptedNames:
      kind: Test
      listKind: TestList
      plural: tests
      singular: test
    conditions:
    - lastTransitionTime: "2020-03-23T05:36:07Z"
      message: no conflicts found
      reason: NoConflicts
      status: "True"
      type: NamesAccepted
    - lastTransitionTime: "2020-03-23T05:36:12Z"
      message: the initial names have been accepted
      reason: InitialNamesAccepted
      status: "True"
      type: Established
    - lastTransitionTime: "2020-03-23T05:36:07Z"
      message: 'spec.validation.openAPIV3Schema.type: Required value: must not be empty
        at the root'
      reason: Violations
      status: "True"
      type: NonStructuralSchema
    storedVersions:
    - v1alpha1
  ```
</details>

File: k8s_handle/k8s/diff.py (hunk 1 above)
Comment (rvadim) [reply k8s_handle/k8s/diff.py:19]: Unfortunatly we don't known the namespace at the diff phase due to we apply namespace in api calls(patch_namespaced_<object>)

**Comment by rvadim** [comment]
@seleznev looks like CRD works now, can you check it?
```
2020-10-16 22:29:30 INFO:k8s_handle.k8s.diff: Kind: "CustomResourceDefinition", name: "tests.test.example.com"
  apiVersion: apiextensions.k8s.io/v1beta1
  kind: CustomResourceDefinition
  metadata:
    name: tests.test.example.com
  spec:
-   conversion:
-     strategy: None
    group: test.example.com
    names:
      kind: Test
-     listKind: TestList
      plural: tests
-     singular: test
    scope: Namespaced
    validation:
      openAPIV3Schema:
        properties:
          apiVersion:
            description: test description
            type: string
          metadata:
            type: object
    version: v1alpha1
-   versions:
-   - name: v1alpha1
-     served: true
-     storage: true

```

File: k8s_handle/k8s/diff.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+from difflib import ndiff
+from datetime import datetime
+from functools import reduce
+import operator
+import yaml
+from .adapters import Adapter
+from k8s_handle.templating import get_template_contexts
+log = logging.getLogger(__name__)
+
+IGNORE_FIELDS = [
```
Comment (seleznev) [inline k8s_handle/k8s/diff.py:13]: What about `metadata.managedFields` (was added in 1.18)? It looks like it should be ignored as well.

Example:
```
  apiVersion: apiextensions.k8s.io/v1beta1
  kind: CustomResourceDefinition
  metadata:
-   managedFields:
-   - apiVersion: apiextensions.k8s.io/v1beta1
-     manager: Swagger-Codegen
-     operation: Update
-     time: '2020-10-19 07:14:08+00:00'
-   - apiVersion: apiextensions.k8s.io/v1
-     manager: kube-apiserver
-     operation: Update
-     time: '2020-10-19 07:14:13+00:00'
    name: tests.test.example.com
```

**Comment by seleznev** [comment]
> @seleznev looks like CRD works now, can you check it?

Yes, it works fine now.

File: k8s_handle/k8s/diff.py (hunk 2 above)
Comment (rvadim) [reply k8s_handle/k8s/diff.py:13]: Added, please check.

**Review by seleznev — APPROVED** [review]

**Review by furiousassault — APPROVED** [review]
If not to mention codestyle flaws like differently nested code blocks which are not separated by spaces, if-else in `_handler_provision`, and similar nagging, LGTM.

**Pull request merged**